sdks/go/pkg/beam/io/fileio/match.go

```diff
@@ -0,0 +1,181 @@
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fileio provides transforms for matching and reading files.
+package fileio
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/apache/beam/sdks/v2/go/pkg/beam"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/io/filesystem"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
+)
+
+func init() {
+	register.DoFn3x1[context.Context, string, func(FileMetadata), error](&matchFn{})
+	register.Emitter1[FileMetadata]()
+}
+
+// emptyTreatment controls how empty matches of a pattern are treated.
+type emptyTreatment int
+
+const (
+	// emptyAllow allows empty matches.
+	emptyAllow emptyTreatment = iota
+	// emptyDisallow disallows empty matches.
+	emptyDisallow
+	// emptyAllowIfWildcard allows empty matches if the pattern contains a wildcard.
+	emptyAllowIfWildcard
+)
+
+type matchOption struct {
+	EmptyTreatment emptyTreatment
+}
+
+// MatchOptionFn is a function that can be passed to MatchFiles or MatchAll to configure options for
+// matching files.
+type MatchOptionFn func(*matchOption)
+
+// MatchEmptyAllow specifies that empty matches are allowed.
+func MatchEmptyAllow() MatchOptionFn {
+	return func(o *matchOption) {
+		o.EmptyTreatment = emptyAllow
+	}
+}
+
+// MatchEmptyDisallow specifies that empty matches are not allowed.
+func MatchEmptyDisallow() MatchOptionFn {
+	return func(o *matchOption) {
+		o.EmptyTreatment = emptyDisallow
+	}
+}
+
+// MatchFiles finds all files matching the glob pattern and returns a PCollection<FileMetadata> of
+// the matching files. MatchFiles accepts a variadic number of MatchOptionFn that can be used to
+// configure the treatment of empty matches. By default, empty matches are allowed if the pattern
+// contains a wildcard.
+func MatchFiles(s beam.Scope, glob string, opts ...MatchOptionFn) beam.PCollection {
+	s = s.Scope("fileio.MatchFiles")
+
+	filesystem.ValidateScheme(glob)
+	return MatchAll(s, beam.Create(s, glob), opts...)
+}
+
+// MatchAll finds all files matching the glob patterns given by the incoming PCollection<string> and
+// returns a PCollection<FileMetadata> of the matching files. MatchAll accepts a variadic number of
+// MatchOptionFn that can be used to configure the treatment of empty matches. By default, empty
+// matches are allowed if the pattern contains a wildcard.
+func MatchAll(s beam.Scope, col beam.PCollection, opts ...MatchOptionFn) beam.PCollection {
+	s = s.Scope("fileio.MatchAll")
+
+	option := &matchOption{
+		EmptyTreatment: emptyAllowIfWildcard,
```

Comment:
I haven't provided any option functions for setting what is already the default value, i.e. there is no `MatchEmptyAllowIfWildcard()`, because it doesn't make much sense? Unless it should be included for clarity/documentation purposes?

Reply:
Okay, fixed now. Feel free to merge if approved